docs/rfc/rfc-019-config-version.md

```diff
@@ -0,0 +1,269 @@
+implies a corresponding config section, but it omits the related detail that
+there was a top-level setting that's been renamed).  The lesson here is not
+that the docs are bad, but to point out that prose is not the most efficient
+format to convey detailed changes like this.
+
+### Concrete Steps
+
+As part of the v0.36 development cycle, we spent some time reverse-engineering
+the configuration changes since the v0.34 release and built an experimental
+command-line tool called [`confix`][confix], whose job it is to automatically
+update the settings in a `config.toml` file to the latest version.  We also
+backported a version of this tool into the v0.35.x branch at release v0.35.4.
+
+This tool should work fine for configuration files created by Tendermint v0.34
+and later, but does not (yet) know how to handle changes from prior versions of
+Tendermint. Part of the difficulty for older versions is simply logistical: To
+figure out which changes to apply, we need to understand something about the
+version that made the file, as well as the version we're converting it to.
+
+> **Discussion point:** In the future we might want to consider incorporating
+> this into the node CLI directly, but we're keeping it separate for now until
+> we can get some feedback from operators.
+
+For the experiment, we handled this by carefully searching the history of
+config format changes for shibboleths to bound the version: For exampole, the
+`[fastsync]` section was added in Tendermint v0.32 and renamed `[blocksync]` in
+Tendermint v0.35. So if we see a `[fastsync]` section, we have some confidence
+that the file was created by v0.32, v0.33, or v0.34.
+
+But such signals are delicate: The `[blocksync]` section was removed in v0.36,
+so if we do not find `[fastsync]`, we cannot conclude from that alone that the
+file is from v0.31 or earlier -- we have to look for corroborating details.
+While such "sniffing" tactics are fine for an experiment, they aren't as robust
+as we might like.
+
+This is especially relevant for configuration files that may have already been
+manually upgraded across several versions by the time we are asked to update
+them again.  Another related concern is that we'd like to make sure conversion
+is idempotent, so that it would be safe to rerun the tool over an
+already-converted file without breaking anything.
+
+### Config Versioning
+
+One obvious tactic we could use for future releases is add a version marker to
+the config file. This would give tools like `confix` (and the node itself) a
+way to calibrate their expectations. Rather than being a version for the file
+itself, however, this version marker would indicate which version of Tendermint
+is needed to read the file.
+
+Provisionally, this might look something like:
+
+```toml
+# THe minimum version of Tendermint compatible with the contents of
+# this configuration file.
+config-version = 'v0.35'
```

Comment:
I kind of want this to be an integer, but I don't think it's that important

Reply:
I'd be fine with that. In my mental conversation my model of you incorrectly predicted you would prefer a string. 😂

Reply:
This seems like a great idea, and many configuration file formats seem to make use of this approach of specifying the config file version within the config file itself.

Although one of the drawbacks of using a simple integer version value for the config file is that you now need to maintain a compatibility matrix between config file versions and versions of Tendermint, which seems tedious to me.

Just using the major/minor Tendermint versions in a string seems more optimal and intuitive.

Reply:
> Although one of the drawbacks of using a simple integer version value for the config file is that you now need to maintain a compatibility matrix between config file versions and versions of Tendermint, which seems tedious to me.

Yeah, I was thinking more of just encoding the version in the integer, e.g., `fmt.Sprintf("1%03d%03d", major, minor)`.

> Just using the major/minor Tendermint versions in a string seems more optimal and intuitive.

That was my original thought, but I think either works. There are good libraries for parsing semver strings.